### PR TITLE
fix: bound RL importance ratios to prevent non-finite gradients

### LINKED ISSUE #2972: RL loss importance ratio can overflow to non-finite gradients
## Summary

The RL loss computes an importance ratio as:

```python
ratio = exp(trainer_logprob - inference_logprob)
```

When rollout logprobs are stale relative to trainer logprobs, this ratio can overflow and produce non-finite loss/gradients.

## Example

A single token is enough:

```python
trainer_logprob = -2.0
inference_logprob = -91.0
log_ratio = trainer_logprob - inference_logprob  # 89.0
ratio = exp(log_ratio)  # inf in float32/bfloat16
```

If the token contributes to the policy-gradient term, the loss can become `inf` or `-inf`. If the token is masked, `0 * inf` can still become `nan`, so masking does not fully protect the step.

## Impact

Before the optimizer step, the non-finite value can flow through backward into gradients. Without a finite-loss or finite-gradient guard, `optimizer.step()` can apply non-finite gradients to model weights or optimizer state.

This is especially reachable in async/off-policy RL, where the trainer policy and rollout-time inference policy can drift.

## Expected behavior

The policy-gradient importance ratio should be bounded before it multiplies token advantages.

`mismatch_kl` should remain useful as an off-policy diagnostic, but its computation should be numerically safe so monitoring does not receive `inf`/`nan`.

## Proposed fix

- Clip the policy-gradient importance ratio with a configurable cap.
- Keep diagnostic mismatch-KL separate from the clipped training ratio.
- Add a finite-gradient guard before `optimizer.step()`, including paths where gradient clipping is disabled.
- Add CPU tests for overflow, masked-token overflow, IPO/ref-KL paths, config validation, and clipping metrics.

### PR DESCRIPTION
**Pull request by morluto**
## Summary

Bounds RL importance ratios so stale async rollouts cannot produce non-finite policy-gradient losses or let non-finite gradients reach `optimizer.step()`.

For highly off-policy tokens, the policy-gradient term now uses a capped score-function weight instead of an unbounded `exp(trainer_logprob - inference_logprob)`. This intentionally bounds the gradient contribution of those tokens without dropping their gradient entirely.

- Adds `importance_ratio_max` with default `20.0` to the default and IPO RL loss configs.
- Uses a detached, straight-through truncated-importance-sampling surrogate in default, IPO, and ref-KL loss paths. Its forward value is the capped ratio and its gradient is `min(ratio, cap) * advantage * grad(log_policy)`.
- Evaluates `mismatch_kl` diagnostics in float32 with the log-ratio saturated to `[-20, 20]`, keeping both individual values and large token reductions finite.
- Keeps raw log-ratios and corresponding importance ratios in token exports; non-finite ratios serialize as `null`.
- Logs `importance_ratio_clipped` metrics and updates W&B filtering.
- Aborts every rank before `optimizer.step()` when any local gradient shard is non-finite, including when DeepEP disables gradient clipping; ranks with no local gradients still participate in the check.

Fixes #2972.

## Context

RL training computes:

```python
ratio = exp(trainer_logprob - inference_logprob)
```

With async training, rollout logprobs can be stale. A single token can overflow the ratio:

```python
trainer_logprob = -2.0
inference_logprob = -91.0
log_ratio = 89.0
```

In float32/bfloat16, `exp(89)` overflows to `inf`. If the token contributes to the policy-gradient term, the loss can become `inf` or `-inf`; if the token is masked, `0 * inf` can still become `nan`.

Before this PR, that non-finite value could flow through backward and into `optimizer.step()`.

## Fix

This separates three concerns:

- **Training weight:** clipped by `importance_ratio_max` before policy-gradient use.
- **Gradient semantics:** the clipped ratio is detached and applied through a straight-through score-function surrogate. Differentiating through `clamp` directly would give tokens above the cap zero gradient; the surrogate instead gives them the intended bounded gradient.
- **Diagnostic mismatch KL:** computed in float32 from a log-ratio saturated to `[-20, 20]`. This keeps aggregate metrics finite without changing the raw log-ratio used by the squared-log-ratio KL loss. Token exports retain the raw log-ratio and raw importance ratio, serializing non-finite ratios as `null`.

The optimizer sees bounded gradients, monitoring exposes clipped-token frequency and off-policy drift, and the final non-finite-gradient guard remains a second line of defense.

## Validation

Latest revision:

- `tests/unit/train/rl/test_loss_numerics.py` — 10 passed, including capped/unclipped gradient values, masked zero-gradient behavior, default/IPO/ref-KL coverage, raw token-export ratio semantics, float16 inputs, aggregate diagnostic finiteness, and a two-rank Gloo/DTensor case where one rank has an `inf` shard and the other has no gradient.
- Importance-ratio config tests — 9 passed.
- Ruff format/check and `git diff --check` — passed.
- Structured Codex autoreview — no accepted/actionable findings.

Original branch validation:

- `uv run pytest tests/unit/train/rl/test_loss_numerics.py tests/unit/test_configs.py tests/unit/orchestrator/test_advantage.py tests/unit/orchestrator/test_metrics.py -q`
- `uv run ruff check src/prime_rl/trainer/rl/train.py src/prime_rl/trainer/rl/loss.py tests/unit/train/rl/test_loss_numerics.py`

---

> [!NOTE]
> **Medium Risk**
> Changes core RL policy-gradient math and optimizer pre-step behavior; bounded by new tests and defaults, but training dynamics on highly off-policy data will differ from unbounded ratios.
> 
> **Overview**
> Caps **policy-gradient** importance weights when async rollouts make `exp(trainer_logprob - inference_logprob)` explode, and adds a distributed guard so bad gradients never reach `optimizer.step()`.
> 
> **RL loss:** Default and IPO configs gain `importance_ratio_max` (default **20**). Default, IPO, and ref-KL paths use a **detached straight-through surrogate** so forward loss uses the capped ratio but gradients follow truncated importance sampling (`min(ratio, cap) * advantage * ∇log π`), not a hard `clamp` that zeroes clipped tokens. **KL** still uses the **raw** log-ratio; **`mismatch_kl`** diagnostics are computed in float32 with log-ratio saturated to **±20** so metrics stay finite in float16. New **`importance_ratio_clipped`** metrics are logged (W&B filtering updated).
> 
> **Trainer:** `clip_grad_norm_` uses `error_if_nonfinite=True`; when clipping is off (e.g. DeepEP), **`raise_if_nonfinite_gradients`** all-reduces finiteness across ranks before the optimizer step.
> 
> **Docs/tests:** `algorithms.md` documents the surrogate and diagnostic split; `test_loss_numerics.py` covers clipping, gradients, mismatch KL, token export (`null` for non-finite ratios), and multi-rank non-finite grad detection.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d96aba41bce678a72deead828b6c637c3855adc4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by morluto** [comment]
Revised in `8a78bc501` after auditing the clipping gradient semantics.

The previous `exp(log_ratio.clamp(...))` produced zero gradient above the cap. The updated implementation uses a detached straight-through score-function surrogate, preserving the capped forward value while yielding the intended truncated-IS gradient.

Also moved bounded `mismatch_kl` diagnostics to float32 so float16 values and multi-token reductions remain finite.

Added regression coverage for default, IPO, ref-KL, masked/unclipped gradients, and aggregate diagnostic finiteness. Final structured autoreview is clean.

File: src/prime_rl/trainer/rl/train.py [hunk 1]
```diff
@@ -70,6 +70,29 @@
...
+        if param.grad is not None:
+            gradients_are_finite = torch.minimum(
+                gradients_are_finite,
+                torch.isfinite(param.grad).to(torch.int32).amin(),
+            )
+
+    if dist.is_initialized():
+        dist.all_reduce(gradients_are_finite, op=dist.ReduceOp.MIN)
+    if not bool(gradients_are_finite.item()):
+        raise RuntimeError("Non-finite gradients detected before optimizer.step(); aborting to avoid corrupting weights.")
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/prime_rl/trainer/rl/train.py]: Fixed in 30603ab56.

The guard now:
- converts DTensor gradients to their local shards through the existing `_to_local_tensor` path;
- initializes a finite sentinel even when a rank has no local gradients, so no rank returns before the collective;
- runs the same world `MIN` reduction on every initialized rank and raises everywhere if any shard is non-finite.

Added a real two-rank Gloo/DTensor regression: rank 0 has an `inf` gradient shard, rank 1 has `grad=None`, and both ranks raise without desynchronizing. The full numerics file passes (9 tests), Ruff is clean, and structured autoreview reported no actionable findings.

File: src/prime_rl/trainer/rl/loss.py [hunk 2]
```diff
@@ -97,12 +101,31 @@ def _safe_mean(values: Tensor, mask: Tensor) -> Tensor:
...
 ) -> tuple[Tensor, Tensor, Tensor]:
     log_importance_ratio = trainer_logprobs - inference_logprobs
-    importance_ratio = torch.exp(log_importance_ratio)
-    mismatch_kl = importance_ratio - log_importance_ratio - 1
+    safe_log_importance_ratio = log_importance_ratio.float().clamp(
+        min=-MISMATCH_KL_LOG_RATIO_LIMIT,
+        max=MISMATCH_KL_LOG_RATIO_LIMIT,
+    )
+    importance_ratio = torch.exp(safe_log_importance_ratio)
+    mismatch_kl = importance_ratio - safe_log_importance_ratio - 1
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/prime_rl/trainer/rl/loss.py]: Fixed in 5aa270612.

`importance_ratio` is again computed from the raw `log_importance_ratio`; only the separately named `mismatch_kl` diagnostic uses the saturated float32 log-ratio. Token export now preserves `importance_ratio == exp(log_importance_ratio)` for finite values and serializes overflow as `null`.

I also corrected the tests that had encoded the saturated-ratio behavior and added an export-path regression covering a finite raw ratio and an overflow. The focused numerics suite passes (10 tests), the importance-ratio config tests pass (9 tests), and Ruff plus `git diff --check` are clean.
